nimbus/src/enrollment.rs

```diff
@@ -20,28 +23,300 @@ pub enum EnrolledReason {
     OptIn,     // Explicit opt-in.
 }
 
+// These are types we use internally for managing non-enrollments.
+#[derive(Deserialize, Serialize, Debug, Clone, Hash, Eq, PartialEq)]
+pub enum NotEnrolledReason {
+    OptOut,      // The user opted-out of experiments before we ever got enrolled to this one.
+    NotSelected, // The evaluator bucketing did not choose us.
+    NotTargeted, // We are not being target for this experiment.
+    NotRunning,  // The experiment enrollment is paused.
+}
+
+// These are types we use internally for managing disqualifications.
+#[derive(Deserialize, Serialize, Debug, Clone, Hash, Eq, PartialEq)]
+pub enum DisqualifiedReason {
+    OptOut,    // The user opted-out from this experiment or experiments in general.
+    Targeting, // The targeting has changed for an experiment.
+}
+
 // Every experiment has an ExperimentEnrollment, even when we aren't enrolled.
 #[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
 pub struct ExperimentEnrollment {
     pub slug: String,
     pub status: EnrollmentStatus,
 }
 
+pub struct EnrollmentChangeEvent {
+    pub experiment_slug: String,
+    pub branch_slug: String,
+    pub enrollment_id: String,
+    pub reason: Option<String>,
```

Comment:
`reason` as a string seems odd - there are 2 arbitrary strings used in cases that don't make total sense to me - ie, why not a non-optional enum?